### PR TITLE
[MachO LLD] Respect -all_load with --start-lib --end-lib style archives

### PR DESCRIPTION
**Pull request by NuriAmari**
The -all_load flag is intended to force the linker to load all lazy members, but doesn't do so if the archive is specified with --start-lib, --end-lib flags. This patch makes it so that -all_load also applies in this case.

### CODE REVIEW

File: lld/MachO/Driver.cpp [hunk 1]
```diff
@@ -1169,7 +1169,8 @@ static void createFiles(const InputArgList &args) {
 
     switch (opt.getID()) {
     case OPT_INPUT:
-      addFile(rerootPath(arg->getValue()), LoadType::CommandLine, isLazy);
+      addFile(rerootPath(arg->getValue()), LoadType::CommandLine,
+              isLazy && !config->allLoad);
```
Comment (kyulee-com) [inline lld/MachO/Driver.cpp:1173]: It's a bit confusing to override `isLazy` for each case. Instead, can we simply not set `isLazy` below in `case OPT_start_lib:` with `config->allLoad` ?

Comment (kyulee-com) [reply lld/MachO/Driver.cpp:1173]: Got it. Using another variable seems clearer to me.
```
case OPT_start_lib:
  if (inLib)
    error("nested --start-lib");
  inLib = true;
  if (!config->allLoad)
    isLazy = true;
  break;
case OPT_end_lib:
  if (!inLib)
    error("stray --end-lib");
  inLib = false;
  if (!config->allLoad)
    isLazy = false;
  break;
```

Comment (alx32) [reply lld/MachO/Driver.cpp:1173]: What if adding on line ~278:
```
bool loadLazily = isLazy && !config->allLoad;
```
And updating lines 370 + 380 to use `loadLazily` ?

Comment (NuriAmari) [reply lld/MachO/Driver.cpp]: Added a combination of the two, I like having an `inLib` variable to make it clear there's a difference between being inside `--start-lib` and actually loading lazily.

**Comment by MaskRay** [comment]
Mach-O -all_load is like a global variant of ELF --whole-archive --no-whole-archive.
For ELF, we have rejected the proposal to treat `--whole-archive --start-lib a.o --end-lib --no-whole-archive` as `.o`
because there isn't a clear signal for --whole-archive to override --start-lib and we stick with the simplest implementation.

For Mach-O, it might make sense for the non-positional -all_load to override --start-lib.
But I expect more reasoning behind this change.

**Review by kyulee-com — APPROVED** [review]
lgtm. I believe this is required to support distributed thinlto (d-thinlto) by specifying each archive member explicitly using `--start-lib/--end-lib`, instead of passing the archive file directly.
Btw, I'm curious how we handle the `-force_load` case (`case OPT_force_load`) for d-thinlto.

**Comment by NuriAmari** [comment]
> For Mach-O, it might make sense for the non-positional -all_load to override --start-lib.
But I expect more reasoning behind this change.

I think the flag being global makes a difference.

Since it is global, a user can supply the flag to a build system and expect it to work without inspecting the rest of the linker invocation. The same is not true of --whole-archive. The user already needs to see the entire invocation to position the —whole-archive / —no-whole-archive flags correctly, at which point they can just remove —start-lib / —end-lib as required. —We use —start-lib / —end-lib for distributed thin-lto. Users should be able to safely switch between thin-lto and distributed thin-lto without having to audit the custom flags they’ve provided (within reason). 

To accomplish this without this change, the build system would need to inspect each linker invocation and remove —start-lib / —end-lib if -all_load is passed. That seems needlessly complex. It seems to me for a global flag, the behavior of a real archive, and —start-lib —end-lib should be the same.

> Btw, I'm curious how we handle the -force_load case (case OPT_force_load) for d-thinlto.

That is an attribute of the library being consumed (link_whole). Libraries with this attribute are passed to the linker without `--start-lib / --end-lib`.

File: lld/MachO/Driver.cpp [hunk 2]
```diff
@@ -1162,6 +1162,8 @@ static void createFiles(const InputArgList &args) {
   // This loop should be reserved for options whose exact ordering matters.
   // Other options should be handled via filtered() and/or getLastArg().
   bool isLazy = false;
+  bool inLib = false; // If we've processed an opening --start-lib, without a
```
Comment (MaskRay) [inline lld/MachO/Driver.cpp:1165]: When an inline comment span 2 lines, we make it to its own line.

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
> > For Mach-O, it might make sense for the non-positional -all_load to override --start-lib.
> > But I expect more reasoning behind this change.
> 
> I think the flag being global makes a difference.
> 
> Since it is global, a user can supply the flag to a build system and expect it to work without inspecting the rest of the linker invocation. The same is not true of --whole-archive. The user already needs to see the entire invocation to position the —whole-archive / —no-whole-archive flags correctly, at which point they can just remove —start-lib / —end-lib as required. —We use —start-lib / —end-lib for distributed thin-lto. Users should be able to safely switch between thin-lto and distributed thin-lto without having to audit the custom flags they’ve provided (within reason).
> 
> To accomplish this without this change, the build system would need to inspect each linker invocation and remove —start-lib / —end-lib if -all_load is passed. That seems needlessly complex. It seems to me for a global flag, the behavior of a real archive, and —start-lib —end-lib should be the same.
> 
> > Btw, I'm curious how we handle the -force_load case (case OPT_force_load) for d-thinlto.
> 
> That is an attribute of the library being consumed (link_whole). Libraries with this attribute are passed to the linker without `--start-lib / --end-lib`.

Yes, I agree with the reasoning. It's worth capturing some of the thinking into the commit message (first comment).

**Review by alx32 — APPROVED** [review]

**Pull request merged**